File: windows/Gutenkern.Core/TokenGlue.cs
using System.Globalization;
using System.Text;

namespace Gutenkern;

public static class TokenGlue
{
    public const char Joiner = '\u2060';
    public const char ViewSlash = '\u29F8';
    public const char AsciiSlash = '/';

    public static string Apply(string text)
    {
        if (string.IsNullOrEmpty(text))
        {
            return text;
        }

        var enumerator = StringInfo.GetTextElementEnumerator(text);
        var builder = new StringBuilder();
        var first = true;
        while (enumerator.MoveNext())
        {
            if (!first)
            {
                builder.Append(Joiner);
            }

            var element = enumerator.GetTextElement();
            builder.Append(element == "/" ? ViewSlash.ToString() : element);
            first = false;
        }

        return builder.ToString();
    }

    public static string Strip(string text)
    {
        if (string.IsNullOrEmpty(text))
        {
            return text;
        }

        return text.Replace(Joiner.ToString(), "").Replace(ViewSlash, AsciiSlash);
    }

    public static string Clean(string text) => Strip(text);

    public static string ViewText(
        string layoutText,
        IReadOnlyList<ResultToken> tokens,
        Func<ResultToken, bool>? shouldGlue = null)
    {
        shouldGlue ??= static _ => true;
        var builder = new StringBuilder();
        var cursor = 0;
        foreach (var token in tokens)
        {
            if (token.Utf16Start > cursor)
            {
                builder.Append(layoutText[cursor..token.Utf16Start]);
            }

            builder.Append(shouldGlue(token) ? Apply(token.Display) : token.Display);
            cursor = token.Utf16End;
        }

        if (cursor < layoutText.Length)
        {
            builder.Append(layoutText[cursor..]);
        }

        return builder.ToString();
    }

    public static int LayoutIndex(string view, int viewIndex)
    {
        var end = Math.Clamp(viewIndex, 0, view.Length);
        var layout = 0;
        for (var index = 0; index < end; index++)
        {
            if (view[index] != Joiner)
            {
                layout++;
            }
        }

        return layout;
    }

    public static int ViewIndex(string view, int layoutIndex)
    {
        var layout = 0;
        for (var index = 0; index < view.Length; index++)
        {
            if (view[index] == Joiner)
            {
                continue;
            }

            if (layout >= layoutIndex)
            {
                return index;
            }

            layout++;
        }

        return view.Length;
    }
}

File: macos/Gutenkern/Sources/GutenkernCore/TokenGlue.swift
import Foundation

public enum TokenGlue {
    public static let joiner = "\u{2060}"
    public static let joinerUTF16: unichar = 0x2060
    public static let viewSlash: Character = "\u{29F8}"
    public static let viewSlashUTF16: unichar = 0x29F8
    public static let asciiSlash: Character = "/"

    public static func apply(_ text: String) -> String {
        guard !text.isEmpty else {
            return text
        }
        var result = ""
        var first = true
        for cluster in text {
            if !first {
                result += joiner
            }
            result.append(cluster == asciiSlash ? viewSlash : cluster)
            first = false
        }
        return result
    }

    public static func strip(_ text: String) -> String {
        text.replacingOccurrences(of: joiner, with: "")
            .replacingOccurrences(of: String(viewSlash), with: String(asciiSlash))
    }

    public static func clean(_ text: String) -> String {
        strip(text)
    }

    public static func viewText(
        layoutText: String,
        tokens: [ResultToken],
        shouldGlue: (ResultToken) -> Bool = { _ in true }
    ) -> String {
        var result = ""
        var cursor = 0
        let ns = layoutText as NSString
        for token in tokens {
            if token.utf16Start > cursor {
                result += ns.substring(
                    with: NSRange(location: cursor, length: token.utf16Start - cursor)
                )
            }
            result += shouldGlue(token) ? apply(token.display) : token.display
            cursor = token.utf16End
        }
        if cursor < ns.length {
            result += ns.substring(from: cursor)
        }
        return result
    }

    public static func layoutIndex(fromView viewIndex: Int, in view: String) -> Int {
        let ns = view as NSString
        let end = min(max(viewIndex, 0), ns.length)
        var layout = 0
        var index = 0
        while index < end {
            if ns.character(at: index) != joinerUTF16 {
                layout += 1
            }
            index += 1
        }
        return layout
    }

    public static func viewIndex(fromLayout layoutIndex: Int, in view: String) -> Int {
        let ns = view as NSString
        var layout = 0
        var index = 0
        while index < ns.length {
            if ns.character(at: index) == joinerUTF16 {
                index += 1
                continue
            }
            if layout >= layoutIndex {
                return index
            }
            layout += 1
            index += 1
        }
        return ns.length
    }

    public static func viewRange(layoutStart: Int, layoutLength: Int, in view: String) -> NSRange {
        let start = viewIndex(fromLayout: layoutStart, in: view)
        let end = viewIndex(fromLayout: layoutStart + layoutLength, in: view)
        return NSRange(location: start, length: max(0, end - start))
    }
}
